### ISSUE TITLE
Allow units of measure in sequence ranges

### ISSUE THREAD

**Issue by marklam** [fsharp/fslang-suggestions#535]
# Allow units of measure in sequence ranges

I propose we extend sequence expressions to handle ranges involving units of measure. Currently we cannot do the following:
```
let maximumDistance = 3<km>
let minimumDistance = 1<km>
let trials = [ minimumDistance .. maximumDistance ]
```

The existing way of approaching this problem in F# is to generate a range from bare integers and then map a multiplication to add the dimension 

## Pros and Cons 

The advantages of making this adjustment to F# are
- An improvement to the safety of code by removing error potential in dropping the units and then adding them back
- More concise, readable code

The disadvantages of making this adjustment to F# are
- None that I can think of

## Extra informtion

Estimated cost (XS, S, M, L, XL, XXL): 
M? (because of the added complication of ``min .. step .. max``)

Related suggestions: (put links to reated suggestions here)
[Allow units of measure in for loops #534 ](https://github.com/fsharp/fslang-suggestions/issues/534)

## Affadavit (must be submitted)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I would be willing to help test this
* [ ] I or my company would be willing to help crowdfund F# Software Foundation members to work on this

**Comment by gmpl** [comment]
Looks like the problem is that units of measure don't support the <code>GenericOne</code> operator, possibly because being a measure, which is basically a multiplication, 'zero' is an absolute value but 'one' is relative to the measure.

Since the ``(..)`` operator uses the ``GenericOne`` it doesn't work, of course you can write this:

    let trials = [ minimumDistance .. 1<km> .. maximumDistance ]

Alternatively, or if this suggestion doesn't get approved, you can suggest to add ``GenericOne`` to a library that offers language extensions, like [FSharpPlus](https://github.com/gmpl/FSharpPlus) which will redefine some generic operators, like ``(..)`` so the code you posted would work, but will be interesting to understand the reason why it wasn't added to F# in first place.

**Comment by marklam** [comment]
Cool, I didn't know that adding measure to the steps would work!

**Comment by dsyme** [comment]
@gmpl You are correct about the original reasons for not allowing ``GenericOne`` on a generic unit

However there's no reason not to allow it on non-generic units.  So we should really just approve the suggestion I think and find a way to make it happen

**Comment by charlesroddie** [comment]
These extensions are being considered rather ad-hoc in order to make certain patterns shorter, irrespective of whether there is a correspondence to mathematical/physical units and whether the brevity gives a false sense of simplicity.

1. `int<'m>` does not have anything like a mathematical One. If there were a multiplicative identity in `int<X>`, call it O, then O*O=O, but O*O is in `int<'m^2>` and O is in `int<'M>` so this is impossible. So GenericOne should not exist for `int<'m>`.
2. The existence of a multiplicative identity does not imply the existence of a natural successor function. Natural successor functions exist in integers and natural numbers but not in real numbers, while multiplicative identities exist in all of these. In this regard the fact that  `for x = 1. to 3.` currently doesn't compile is correct and the fact that `[1. .. 3.]` currently compiles is a mistake.
3. `int<'m>` exists in F# but not in regular use of units, and the semantics are unclear. `integers<'m>`  is more like a subset of `reals<'m>` than than a projection of `integers`. The code `[1<km> .. 3<km>]`, means "start with 1km, notice that we are only dealing with a set where valid items are multiples of 1km, find the nearest greater valid item in this set", which is conceptually more complex than `[1..3] |> List.map LanguagePrimitives.Int32WithMeasure<km>`, which is nearer to the mathematical formulation `{ (f i)km : i in Z }`, where `f` is the natural injection from integers to reals.
